runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/FileWatcherTask.java

```diff
@@ -0,0 +1,147 @@
+        }
+        catch (IOException ex)
+        {
+            rethrowUnchecked(ex);
+        }
+
+    }
+
+    @Override
+    public boolean run()
+    {
+        try
+        {
+            while (true)
+            {
+                try
+                {
+                    final WatchKey key = watchService.take();
+
+                    for (WatchEvent<?> event : key.pollEvents())
+                    {
+                        final Path changed = ((Path) key.watchable()).resolve((Path) event.context());
+                        if (configHashes.containsKey(changed))
+                        {
+                            String configText = readConfigText(configURLs.get(changed));
+                            byte[] newConfigHash = computeHash(configText);
+                            if (!Arrays.equals(configHashes.get(changed), newConfigHash))
+                            {
+                                configHashes.put(changed, newConfigHash);
+                                configChangeListener.accept(configURLs.get(changed), configText);
+                            }
+                        }
+                    }
+                    key.reset();
+                }
+                catch (InterruptedException ex)
+                {
+                    watchService.close();
+                    break;
+                }
+            }
+        }
+        catch (IOException ex)
+        {
+            rethrowUnchecked(ex);
+        }
+        return true;
+    }
+
+    @Override
+    protected void doInitialConfiguration(URL configURL)
+    {
+        Path configPath = Paths.get(new File(configURL.getPath()).getAbsolutePath());
+        configURLs.put(configPath, configURL);
+        String configText = readConfigText(configURL);
+        configHashes.put(configPath, computeHash(configText));
+
+        try
+        {
+            configPath.getParent().register(watchService, ENTRY_MODIFY, ENTRY_CREATE, ENTRY_DELETE);
+        }
+        catch (IOException ignored)
+        {
+        }
+
+        configChangeListener.accept(configURL, configText);
+    }
+
+    private String readConfigText(URL configURL)
+    {
+        String configText;
+        if (configURL == null)
+        {
+            configText = CONFIG_TEXT_DEFAULT;
+        }
+        else
+        {
+            try
+            {
+                URLConnection connection = configURL.openConnection();
+                try (InputStream input = connection.getInputStream())
+                {
+                    configText = new String(input.readAllBytes(), UTF_8);
+                }
+            }
+            catch (IOException ex)
+            {
+                configText = CONFIG_TEXT_DEFAULT;
+            }
+        }
+        return configText;
+    }
+
+    private byte[] computeHash(String configText)
+    {
+        byte[] hash = new byte[0];
+        try
+        {
+            MessageDigest md5Digest = MessageDigest.getInstance("MD5");
```

Comment:
Call this `md5` and hoist to constructor, instead of reinstantiating for each call to `computeHash`.

This might require creating a private static method to handle the exception, calling from constructor to initialize and return new message digest instance.